File: mobile/ios/packages/town-crier-domain/Sources/Protocols/UserProfileRepository.swift
import Foundation

/// Port for managing the user's server-side profile.
///
/// Maps to the `/v1/me` API endpoints:
/// - `create()` -> `POST /v1/me`
/// - `fetch()` -> `GET /v1/me`
/// - `update(...)` -> `PATCH /v1/me`
/// - `delete()` -> `DELETE /v1/me`
/// - `exportData()` -> `GET /v1/me/data`
public protocol UserProfileRepository: Sendable {
  /// Creates the user profile on the server. The API reads identity from the JWT.
  func create() async throws -> ServerProfile

  /// Fetches the current user profile. Returns `nil` if no profile exists (404).
  func fetch() async throws -> ServerProfile?

  /// Updates mutable profile preferences.
  ///
  /// `savedDecisionPush` and `savedDecisionEmail` default to `true` so existing
  /// callers that only manage digest preferences keep their behaviour without
  /// accidentally toggling saved-application notifications off.
  func update(
    pushEnabled: Bool,
    digestDay: DayOfWeek,
    emailDigestEnabled: Bool,
    savedDecisionPush: Bool,
    savedDecisionEmail: Bool
  ) async throws -> ServerProfile

  /// Deletes the user profile and cascades (removes all user data server-side).
  func delete() async throws

  /// Fetches the full GDPR data export as raw JSON bytes.
  ///
  /// The export is opaque to the client: the server bytes are returned as-is so
  /// the user can save or share a machine-readable copy of all their data. The
  /// caller must not decode or re-encode the payload, to keep the export
  /// byte-stable with what the server produced.
  func exportData() async throws -> Data
}

File: mobile/android/domain/src/main/kotlin/uk/towncrierapp/domain/profile/UserProfileRepository.kt
package uk.towncrierapp.domain.profile

/** Port for managing the user's server-side profile. */
public interface UserProfileRepository {
    /**
     * Ensures the server profile exists (creates it on first call, returns
     * the existing one otherwise — the server-side handler is idempotent)
     * and returns it. Called on every signed-in transition, before tier
     * resolution (tc-a6it / #549 ordering). NOTE: the wire response for this
     * endpoint carries only `userId`/`pushEnabled`/`tier` — the returned
     * [ServerProfile]'s notification-preference fields are the server's
     * defaults, not necessarily this user's actual saved values; use
     * [fetchProfile] to read the real preferences (#778).
     */
    public suspend fun ensureProfile(): ServerProfile

    /** Fetches the current profile (`GET /v1/me`), or `null` if none exists yet (404). */
    public suspend fun fetchProfile(): ServerProfile?

    /**
     * Replaces ALL FIVE notification preferences in one call (`PATCH /v1/me`)
     * — the server treats them as a set, so every setter round-trips every
     * field, never a partial update (epic #770 pre-resolved decision).
     */
    public suspend fun updatePreferences(preferences: UserPreferences): ServerProfile

    /**
     * UK GDPR Art. 17 erasure (`DELETE /v1/me`). The server cascades to
     * every per-user store and, last, deletes the Auth0 user — the client
     * NEVER calls Auth0's management API directly.
     */
    public suspend fun deleteAccount()

    /** The full GDPR data export (`GET /v1/me/data`) as opaque, unmodified bytes. */
    public suspend fun exportData(): ByteArray
}
